votor/src/consensus_pool/vote_pool.rs

```diff
@@ -1,129 +1,304 @@
+                        let len = map.len();
+                        match map.entry(nf.block_id) {
+                            Entry::Vacant(map_e) => {
+                                if len == 3 {
+                                    Err(AddVoteError::Slash)
+                                } else {
+                                    map_e.insert(vote);
+                                    Ok(())
+                                }
+                            }
+                            Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                        }
+                    }
+                }
+            }
+            Vote::Skip(_) => {
+                if self.notar.contains_key(&voter) || self.finalize.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.skip.entry(voter) {
+                    Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
+                        Ok(())
+                    }
+                }
+            }
+            Vote::SkipFallback(_) => {
+                if self.finalize.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.skip_fallback.entry(voter) {
+                    Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
+                        Ok(())
+                    }
+                }
+            }
+            Vote::Finalize(_) => {
+                if self.skip.contains_key(&voter) || self.skip_fallback.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                if let Some(map) = self.notar_fallback.get(&voter) {
+                    assert!(!map.is_empty());
+                    return Err(AddVoteError::Slash);
+                }
+                match self.finalize.entry(voter) {
+                    Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
+                        Ok(())
+                    }
+                }
+            }
+        }
     }
 
-    pub(super) fn has_prev_validator_vote(&self, validator_vote_key: &Pubkey) -> bool {
-        self.prev_voted_validators.contains(validator_vote_key)
+    /// Get votes for the corresponding [`VoteType`] and block id.
+    ///
+    // TODO: figure out how to return an iterator here instead which would require `CertificateBuilder::aggregate()` to accept an iterator.
+    // TODO: instead of passing vote_type and block_id, pass in `Vote` which will remove some unwraps below.
+    fn get_votes(&self, vote_type: &VoteType, block_id: Option<&Hash>) -> Vec<VoteMessage> {
+        match vote_type {
+            VoteType::Finalize => self.finalize.values().cloned().collect(),
```

Comment:
Err, we are doing a copy every time now?

Reply:
Yes, as mentioned in the other comment, this is drawback of this PR.  I have added a TODO to fix it and I have a work in progress branch where I plan on fixing it.